File: step_16_dynamic_programming/16.2_1d_dp/P003_house_robber_ii/Solution.java
/**
 * Problem: House Robber II (LeetCode #213)
 * Difficulty: MEDIUM | XP: 25
 *
 * Houses in a circle. Run House Robber twice: skip first OR skip last.
 * All 4 DP approaches: Recursion -> Memo -> Tab -> Space Optimized
 *
 * @author DSA_Nova
 */
import java.util.Arrays;

// ============================================================
// Approach 1: Plain Recursion
// Time: O(2^n) | Space: O(n)
// ============================================================
class RecursiveRobber {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1) return nums[0];

        // Max of: skip last house, skip first house
        return Math.max(
            solveLinear(nums, 0, n - 2),
            solveLinear(nums, 1, n - 1)
        );
    }

    private int solveLinear(int[] nums, int start, int end) {
        return solve(nums, end, start);
    }

    private int solve(int[] nums, int i, int start) {
        if (i < start) return 0;
        if (i == start) return nums[start];

        return Math.max(
            solve(nums, i - 1, start),           // skip house i
            nums[i] + solve(nums, i - 2, start)  // rob house i
        );
    }
}

// ============================================================
// Approach 2: Memoization (Top-Down DP)
// Time: O(n) | Space: O(n)
// ============================================================
class MemoRobber {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1) return nums[0];

        return Math.max(
            solveLinear(nums, 0, n - 2),
            solveLinear(nums, 1, n - 1)
        );
    }

    private int solveLinear(int[] nums, int start, int end) {
        int[] dp = new int[nums.length];
        Arrays.fill(dp, -1);
        return solve(nums, end, start, dp);
    }

    private int solve(int[] nums, int i, int start, int[] dp) {
        if (i < start) return 0;
        if (i == start) return nums[start];
        if (dp[i] != -1) return dp[i];

        dp[i] = Math.max(
            solve(nums, i - 1, start, dp),
            nums[i] + solve(nums, i - 2, start, dp)
        );
        return dp[i];
    }
}

// ============================================================
// Approach 3: Tabulation (Bottom-Up DP)
// Time: O(n) | Space: O(n)
// ============================================================
class TabRobber {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1) return nums[0];

        return Math.max(
            robLinear(nums, 0, n - 2),
            robLinear(nums, 1, n - 1)
        );
    }

    private int robLinear(int[] nums, int start, int end) {
        int len = end - start + 1;
        if (len == 1) return nums[start];

        int[] dp = new int[len];
        dp[0] = nums[start];
        dp[1] = Math.max(nums[start], nums[start + 1]);

        for (int i = 2; i < len; i++) {
            dp[i] = Math.max(dp[i - 1], nums[start + i] + dp[i - 2]);
        }

        return dp[len - 1];
    }
}

// ============================================================
// Approach 4: Space Optimized
// Time: O(n) | Space: O(1)
// ============================================================
class SpaceRobber {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1) return nums[0];

        return Math.max(
            robLinear(nums, 0, n - 2),
            robLinear(nums, 1, n - 1)
        );
    }

    private int robLinear(int[] nums, int start, int end) {
        int prev2 = 0, prev1 = 0;

        for (int i = start; i <= end; i++) {
            int curr = Math.max(prev1, nums[i] + prev2);
            prev2 = prev1;
            prev1 = curr;
        }

        return prev1;
    }
}

// ============================================================
// Main driver
// ============================================================
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== House Robber II ===\n");

        RecursiveRobber rec = new RecursiveRobber();
        MemoRobber memo = new MemoRobber();
        TabRobber tab = new TabRobber();
        SpaceRobber space = new SpaceRobber();

        int[][] houses = {
            {2, 3, 2},
            {1, 2, 3, 1},
            {1, 2, 3},
            {0},
            {5},
            {1, 2},
            {3, 3, 3, 3},
            {1, 100, 1, 100},
        };
        int[] expected = {3, 4, 3, 0, 5, 2, 6, 200};

        for (int t = 0; t < houses.length; t++) {
            int r = rec.rob(houses[t]);
            int m = memo.rob(houses[t]);
            int tb = tab.rob(houses[t]);
            int s = space.rob(houses[t]);

            boolean pass = r == expected[t] && m == expected[t]
                    && tb == expected[t] && s == expected[t];

            System.out.println("nums=" + Arrays.toString(houses[t]));
            System.out.println("  Rec=" + r + " | Memo=" + m + " | Tab=" + tb + " | Space=" + s);
            System.out.println("  Expected=" + expected[t] + " | Pass=" + pass + "\n");
        }
    }
}

File: step_16_dynamic_programming/16.2_1d_dp/P003_house_robber_ii/solution.py
"""
Problem: House Robber II (LeetCode #213)
Difficulty: MEDIUM | XP: 25

Houses in a circle. Run House Robber twice: skip first OR skip last.
All 4 DP approaches: Recursion -> Memo -> Tab -> Space Optimized
"""
from typing import List


# ============================================================
# APPROACH 1: PLAIN RECURSION
# Time: O(2^n) | Space: O(n)
# ============================================================
def rob_recursive(nums: List[int]) -> int:
    """Break circle into two linear problems, solve each recursively."""
    n = len(nums)
    if n == 1:
        return nums[0]

    def solve(i: int, start: int) -> int:
        if i < start:
            return 0
        if i == start:
            return nums[start]
        return max(solve(i - 1, start), nums[i] + solve(i - 2, start))

    return max(solve(n - 2, 0), solve(n - 1, 1))


# ============================================================
# APPROACH 2: MEMOIZATION (Top-Down DP)
# Time: O(n) | Space: O(n)
# ============================================================
def rob_memo(nums: List[int]) -> int:
    """Cache states for each linear sub-problem."""
    n = len(nums)
    if n == 1:
        return nums[0]

    def solve_linear(start: int, end: int) -> int:
        dp = {}

        def solve(i: int) -> int:
            if i < start:
                return 0
            if i == start:
                return nums[start]
            if i in dp:
                return dp[i]
            dp[i] = max(solve(i - 1), nums[i] + solve(i - 2))
            return dp[i]

        return solve(end)

    return max(solve_linear(0, n - 2), solve_linear(1, n - 1))


# ============================================================
# APPROACH 3: TABULATION (Bottom-Up DP)
# Time: O(n) | Space: O(n)
# ============================================================
def rob_tab(nums: List[int]) -> int:
    """Build dp array for each linear sub-problem."""
    n = len(nums)
    if n == 1:
        return nums[0]

    def rob_linear(start: int, end: int) -> int:
        length = end - start + 1
        if length == 1:
            return nums[start]

        dp = [0] * length
        dp[0] = nums[start]
        dp[1] = max(nums[start], nums[start + 1])

        for i in range(2, length):
            dp[i] = max(dp[i - 1], nums[start + i] + dp[i - 2])

        return dp[length - 1]

    return max(rob_linear(0, n - 2), rob_linear(1, n - 1))


# ============================================================
# APPROACH 4: SPACE OPTIMIZED
# Time: O(n) | Space: O(1)
# ============================================================
def rob_space(nums: List[int]) -> int:
    """Use two variables per linear sub-problem. O(1) space."""
    n = len(nums)
    if n == 1:
        return nums[0]

    def rob_linear(start: int, end: int) -> int:
        prev2, prev1 = 0, 0
        for i in range(start, end + 1):
            curr = max(prev1, nums[i] + prev2)
            prev2 = prev1
            prev1 = curr
        return prev1

    return max(rob_linear(0, n - 2), rob_linear(1, n - 1))


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== House Robber II ===\n")

    test_cases = [
        ([2, 3, 2], 3),
        ([1, 2, 3, 1], 4),
        ([1, 2, 3], 3),
        ([0], 0),
        ([5], 5),
        ([1, 2], 2),
        ([3, 3, 3, 3], 6),
        ([1, 100, 1, 100], 200),
    ]

    for nums, expected in test_cases:
        r = rob_recursive(nums)
        m = rob_memo(nums)
        t = rob_tab(nums)
        s = rob_space(nums)

        passes = r == expected and m == expected and t == expected and s == expected
        print(f"nums={nums}")
        print(f"  Rec={r} | Memo={m} | Tab={t} | Space={s}")
        print(f"  Expected={expected} | Pass={passes}\n")
